File: samples/client/3_0_3_unit_test/java/src/test/java/org/openapijsonschematools/client/components/schemas/UniqueitemsValidationTest.java
package org.openapijsonschematools.client.components.schemas;

import org.junit.Assert;
import org.junit.Test;
import org.openapijsonschematools.client.configurations.JsonSchemaKeywordFlags;
import org.openapijsonschematools.client.configurations.SchemaConfiguration;
import org.openapijsonschematools.client.exceptions.ValidationException;
import org.openapijsonschematools.client.schemas.validation.MapUtils;
import org.checkerframework.checker.nullness.qual.Nullable;

import java.util.Arrays;
import java.util.List;
import java.util.Map;
import java.util.AbstractMap;

public class UniqueitemsValidationTest {
    static final SchemaConfiguration configuration = new SchemaConfiguration(new JsonSchemaKeywordFlags.Builder().format().build());

    @Test
    public void testNonUniqueArrayOfMoreThanTwoIntegersIsInvalidFails() {
        // non-unique array of more than two integers is invalid
        final var schema = UniqueitemsValidation.UniqueitemsValidation1.getInstance();
        try {
            schema.validate(
                Arrays.asList(
                    1,
                    2,
                    1
                ),
                configuration
            );
            throw new RuntimeException("A different exception must be thrown");
        } catch (ValidationException ignored) {
            ;
        }
    }

    @Test
    public void testNonUniqueArrayOfObjectsIsInvalidFails() {
        // non-unique array of objects is invalid
        final var schema = UniqueitemsValidation.UniqueitemsValidation1.getInstance();
        try {
            schema.validate(
                Arrays.asList(
                    MapUtils.makeMap(
                        new AbstractMap.SimpleEntry<String, String>(
                            "foo",
                            "bar"
                        )
                    ),
                    MapUtils.makeMap(
                        new AbstractMap.SimpleEntry<String, String>(
                            "foo",
                            "bar"
                        )
                    )
                ),
                configuration
            );
            throw new RuntimeException("A different exception must be thrown");
        } catch (ValidationException ignored) {
            ;
        }
    }

    @Test
    public void testATrueAndA1AreUniquePasses() throws ValidationException {
        // {\\\&quot;a\\\&quot;: true} and {\\\&quot;a\\\&quot;: 1} are unique
        final var schema = UniqueitemsValidation.UniqueitemsValidation1.getInstance();
        schema.validate(
            Arrays.asList(
                MapUtils.makeMap(
                    new AbstractMap.SimpleEntry<String, Boolean>(
                        "a",
                        true
                    )
                ),
                MapUtils.makeMap(
                    new AbstractMap.SimpleEntry<String, Number>(
                        "a",
                        1
                    )
                )
            ),
            configuration
        );
    }

    @Test
    public void test1AndTrueAreUniquePasses() throws ValidationException {
        // [1] and [true] are unique
        final var schema = UniqueitemsValidation.UniqueitemsValidation1.getInstance();
        schema.validate(
            Arrays.asList(
                Arrays.asList(
                    1
                ),
                Arrays.asList(
                    true
                )
            ),
            configuration
        );
    }

    @Test
    public void testNonUniqueArrayOfIntegersIsInvalidFails() {
        // non-unique array of integers is invalid
        final var schema = UniqueitemsValidation.UniqueitemsValidation1.getInstance();
        try {
            schema.validate(
                Arrays.asList(
                    1,
                    1
                ),
                configuration
            );
            throw new RuntimeException("A different exception must be thrown");
        } catch (ValidationException ignored) {
            ;
        }
    }

    @Test
    public void testNested0AndFalseAreUniquePasses() throws ValidationException {
        // nested [0] and [false] are unique
        final var schema = UniqueitemsValidation.UniqueitemsValidation1.getInstance();
        schema.validate(
            Arrays.asList(
                Arrays.asList(
                    Arrays.asList(
                        0
                    ),
                    "foo"
                ),
                Arrays.asList(
                    Arrays.asList(
                        false
                    ),
                    "foo"
                )
            ),
            configuration
        );
    }

    @Test
    public void testObjectsAreNonUniqueDespiteKeyOrderFails() {
        // objects are non-unique despite key order
        final var schema = UniqueitemsValidation.UniqueitemsValidation1.getInstance();
        try {
            schema.validate(
                Arrays.asList(
                    MapUtils.makeMap(
                        new AbstractMap.SimpleEntry<String, Number>(
                            "a",
                            1
                        ),
                        new AbstractMap.SimpleEntry<String, Number>(
                            "b",
                            2
                        )
                    ),
                    MapUtils.makeMap(
                        new AbstractMap.SimpleEntry<String, Number>(
                            "b",
                            2
                        ),
                        new AbstractMap.SimpleEntry<String, Number>(
                            "a",
                            1
                        )
                    )
                ),
                configuration
            );
            throw new RuntimeException("A different exception must be thrown");
        } catch (ValidationException ignored) {
            ;
        }
    }

    @Test
    public void testNonUniqueArrayOfArraysIsInvalidFails() {
        // non-unique array of arrays is invalid
        final var schema = UniqueitemsValidation.UniqueitemsValidation1.getInstance();
        try {
            schema.validate(
                Arrays.asList(
                    Arrays.asList(
                        "foo"
                    ),
                    Arrays.asList(
                        "foo"
                    )
                ),
                configuration
            );
            throw new RuntimeException("A different exception must be thrown");
        } catch (ValidationException ignored) {
            ;
        }
    }

    @Test
    public void testAFalseAndA0AreUniquePasses() throws ValidationException {
        // {\\\&quot;a\\\&quot;: false} and {\\\&quot;a\\\&quot;: 0} are unique
        final var schema = UniqueitemsValidation.UniqueitemsValidation1.getInstance();
        schema.validate(
            Arrays.asList(
                MapUtils.makeMap(
                    new AbstractMap.SimpleEntry<String, Boolean>(
                        "a",
                        false
                    )
                ),
                MapUtils.makeMap(
                    new AbstractMap.SimpleEntry<String, Number>(
                        "a",
                        0
                    )
                )
            ),
            configuration
        );
    }

    @Test
    public void testNonUniqueArrayOfMoreThanTwoArraysIsInvalidFails() {
        // non-unique array of more than two arrays is invalid
        final var schema = UniqueitemsValidation.UniqueitemsValidation1.getInstance();
        try {
            schema.validate(
                Arrays.asList(
                    Arrays.asList(
                        "foo"
                    ),
                    Arrays.asList(
                        "bar"
                    ),
                    Arrays.asList(
                        "foo"
                    )
                ),
                configuration
            );
            throw new RuntimeException("A different exception must be thrown");
        } catch (ValidationException ignored) {
            ;
        }
    }

    @Test
    public void test0AndFalseAreUniquePasses() throws ValidationException {
        // [0] and [false] are unique
        final var schema = UniqueitemsValidation.UniqueitemsValidation1.getInstance();
        schema.validate(
            Arrays.asList(
                Arrays.asList(
                    0
                ),
                Arrays.asList(
                    false
                )
            ),
            configuration
        );
    }

    @Test
    public void testNonUniqueArrayOfNestedObjectsIsInvalidFails() {
        // non-unique array of nested objects is invalid
        final var schema = UniqueitemsValidation.UniqueitemsValidation1.getInstance();
        try {
            schema.validate(
                Arrays.asList(
                    MapUtils.makeMap(
                        new AbstractMap.SimpleEntry<String, Map<String, ?>>(
                            "foo",
                            MapUtils.makeMap(
                                new AbstractMap.SimpleEntry<String, Map<String, ?>>(
                                    "bar",
                                    MapUtils.makeMap(
                                        new AbstractMap.SimpleEntry<String, Boolean>(
                                            "baz",
                                            true
                                        )
                                    )
                                )
                            )
                        )
                    ),
                    MapUtils.makeMap(
                        new AbstractMap.SimpleEntry<String, Map<String, ?>>(
                            "foo",
                            MapUtils.makeMap(
                                new AbstractMap.SimpleEntry<String, Map<String, ?>>(
                                    "bar",
                                    MapUtils.makeMap(
                                        new AbstractMap.SimpleEntry<String, Boolean>(
                                            "baz",
                                            true
                                        )
                                    )
                                )
                            )
                        )
                    )
                ),
                configuration
            );
            throw new RuntimeException("A different exception must be thrown");
        } catch (ValidationException ignored) {
            ;
        }
    }

    @Test
    public void testNumbersAreUniqueIfMathematicallyUnequalFails() {
        // numbers are unique if mathematically unequal
        final var schema = UniqueitemsValidation.UniqueitemsValidation1.getInstance();
        try {
            schema.validate(
                Arrays.asList(
                    1.0d,
                    1.0d,
                    1
                ),
                configuration
            );
            throw new RuntimeException("A different exception must be thrown");
        } catch (ValidationException ignored) {
            ;
        }
    }

    @Test
    public void testNonUniqueArrayOfStringsIsInvalidFails() {
        // non-unique array of strings is invalid
        final var schema = UniqueitemsValidation.UniqueitemsValidation1.getInstance();
        try {
            schema.validate(
                Arrays.asList(
                    "foo",
                    "bar",
                    "foo"
                ),
                configuration
            );
            throw new RuntimeException("A different exception must be thrown");
        } catch (ValidationException ignored) {
            ;
        }
    }

    @Test
    public void testUniqueArrayOfNestedObjectsIsValidPasses() throws ValidationException {
        // unique array of nested objects is valid
        final var schema = UniqueitemsValidation.UniqueitemsValidation1.getInstance();
        schema.validate(
            Arrays.asList(
                MapUtils.makeMap(
                    new AbstractMap.SimpleEntry<String, Map<String, ?>>(
                        "foo",
                        MapUtils.makeMap(
                            new AbstractMap.SimpleEntry<String, Map<String, ?>>(
                                "bar",
                                MapUtils.makeMap(
                                    new AbstractMap.SimpleEntry<String, Boolean>(
                                        "baz",
                                        true
                                    )
                                )
                            )
                        )
                    )
                ),
                MapUtils.makeMap(
                    new AbstractMap.SimpleEntry<String, Map<String, ?>>(
                        "foo",
                        MapUtils.makeMap(
                            new AbstractMap.SimpleEntry<String, Map<String, ?>>(
                                "bar",
                                MapUtils.makeMap(
                                    new AbstractMap.SimpleEntry<String, Boolean>(
                                        "baz",
                                        false
                                    )
                                )
                            )
                        )
                    )
                )
            ),
            configuration
        );
    }

    @Test
    public void testUniqueArrayOfArraysIsValidPasses() throws ValidationException {
        // unique array of arrays is valid
        final var schema = UniqueitemsValidation.UniqueitemsValidation1.getInstance();
        schema.validate(
            Arrays.asList(
                Arrays.asList(
                    "foo"
                ),
                Arrays.asList(
                    "bar"
                )
            ),
            configuration
        );
    }

    @Test
    public void testTrueIsNotEqualToOnePasses() throws ValidationException {
        // true is not equal to one
        final var schema = UniqueitemsValidation.UniqueitemsValidation1.getInstance();
        schema.validate(
            Arrays.asList(
                1,
                true
            ),
            configuration
        );
    }

    @Test
    public void testNested1AndTrueAreUniquePasses() throws ValidationException {
        // nested [1] and [true] are unique
        final var schema = UniqueitemsValidation.UniqueitemsValidation1.getInstance();
        schema.validate(
            Arrays.asList(
                Arrays.asList(
                    Arrays.asList(
                        1
                    ),
                    "foo"
                ),
                Arrays.asList(
                    Arrays.asList(
                        true
                    ),
                    "foo"
                )
            ),
            configuration
        );
    }

    @Test
    public void testUniqueArrayOfStringsIsValidPasses() throws ValidationException {
        // unique array of strings is valid
        final var schema = UniqueitemsValidation.UniqueitemsValidation1.getInstance();
        schema.validate(
            Arrays.asList(
                "foo",
                "bar",
                "baz"
            ),
            configuration
        );
    }

    @Test
    public void testFalseIsNotEqualToZeroPasses() throws ValidationException {
        // false is not equal to zero
        final var schema = UniqueitemsValidation.UniqueitemsValidation1.getInstance();
        schema.validate(
            Arrays.asList(
                0,
                false
            ),
            configuration
        );
    }

    @Test
    public void testUniqueArrayOfIntegersIsValidPasses() throws ValidationException {
        // unique array of integers is valid
        final var schema = UniqueitemsValidation.UniqueitemsValidation1.getInstance();
        schema.validate(
            Arrays.asList(
                1,
                2
            ),
            configuration
        );
    }

    @Test
    public void testDifferentObjectsAreUniquePasses() throws ValidationException {
        // different objects are unique
        final var schema = UniqueitemsValidation.UniqueitemsValidation1.getInstance();
        schema.validate(
            Arrays.asList(
                MapUtils.makeMap(
                    new AbstractMap.SimpleEntry<String, Number>(
                        "a",
                        1
                    ),
                    new AbstractMap.SimpleEntry<String, Number>(
                        "b",
                        2
                    )
                ),
                MapUtils.makeMap(
                    new AbstractMap.SimpleEntry<String, Number>(
                        "a",
                        2
                    ),
                    new AbstractMap.SimpleEntry<String, Number>(
                        "b",
                        1
                    )
                )
            ),
            configuration
        );
    }

    @Test
    public void testUniqueHeterogeneousTypesAreValidPasses() throws ValidationException {
        // unique heterogeneous types are valid
        final var schema = UniqueitemsValidation.UniqueitemsValidation1.getInstance();
        schema.validate(
            Arrays.asList(
                MapUtils.makeMap(
                ),
                Arrays.asList(
                    1
                ),
                true,
                null,
                1,
                "{}"
            ),
            configuration
        );
    }

    @Test
    public void testUniqueArrayOfObjectsIsValidPasses() throws ValidationException {
        // unique array of objects is valid
        final var schema = UniqueitemsValidation.UniqueitemsValidation1.getInstance();
        schema.validate(
            Arrays.asList(
                MapUtils.makeMap(
                    new AbstractMap.SimpleEntry<String, String>(
                        "foo",
                        "bar"
                    )
                ),
                MapUtils.makeMap(
                    new AbstractMap.SimpleEntry<String, String>(
                        "foo",
                        "baz"
                    )
                )
            ),
            configuration
        );
    }

    @Test
    public void testNonUniqueHeterogeneousTypesAreInvalidFails() {
        // non-unique heterogeneous types are invalid
        final var schema = UniqueitemsValidation.UniqueitemsValidation1.getInstance();
        try {
            schema.validate(
                Arrays.asList(
                    MapUtils.makeMap(
                    ),
                    Arrays.asList(
                        1
                    ),
                    true,
                    null,
                    MapUtils.makeMap(
                    ),
                    1
                ),
                configuration
            );
            throw new RuntimeException("A different exception must be thrown");
        } catch (ValidationException ignored) {
            ;
        }
    }
}

File: samples/client/3_0_3_unit_test/kotlin/src/test/kotlin/org/openapijsonschematools/client/components/schemas/UniqueitemsValidationTest.kt
package org.openapijsonschematools.client.components.schemas

import org.openapijsonschematools.client.configurations.JsonSchemaKeywordFlags
import org.openapijsonschematools.client.configurations.SchemaConfiguration
import org.openapijsonschematools.client.exceptions.ValidationException

import kotlin.test.Test
import kotlin.test.assertFailsWith

class UniqueitemsValidationTest {
    companion object {
        val configuration = SchemaConfiguration(JsonSchemaKeywordFlags.Builder().format().build())
    }

    @Test
    fun testNonUniqueArrayOfMoreThanTwoIntegersIsInvalidFails() {
        // non-unique array of more than two integers is invalid
        val schema = UniqueitemsValidation.UniqueitemsValidation1.getInstance()
        assertFailsWith<ValidationException>(
            block = {
                schema.validate(
                    listOf(
                        1,
                        2,
                        1
                    ),
                    configuration
                )
            }
        )
    }

    @Test
    fun testNonUniqueArrayOfObjectsIsInvalidFails() {
        // non-unique array of objects is invalid
        val schema = UniqueitemsValidation.UniqueitemsValidation1.getInstance()
        assertFailsWith<ValidationException>(
            block = {
                schema.validate(
                    listOf(
                        mapOf(
                            Pair(
                                "foo",
                                "bar"
                            )
                        ),
                        mapOf(
                            Pair(
                                "foo",
                                "bar"
                            )
                        )
                    ),
                    configuration
                )
            }
        )
    }

    @Test
    fun testATrueAndA1AreUniquePasses() {
        // {\\\&quot;a\\\&quot;: true} and {\\\&quot;a\\\&quot;: 1} are unique
        val schema = UniqueitemsValidation.UniqueitemsValidation1.getInstance()
        schema.validate(
            listOf(
                mapOf(
                    Pair(
                        "a",
                        true
                    )
                ),
                mapOf(
                    Pair(
                        "a",
                        1
                    )
                )
            ),
            configuration
        )
    }

    @Test
    fun test1AndTrueAreUniquePasses() {
        // [1] and [true] are unique
        val schema = UniqueitemsValidation.UniqueitemsValidation1.getInstance()
        schema.validate(
            listOf(
                listOf(
                    1
                ),
                listOf(
                    true
                )
            ),
            configuration
        )
    }

    @Test
    fun testNonUniqueArrayOfIntegersIsInvalidFails() {
        // non-unique array of integers is invalid
        val schema = UniqueitemsValidation.UniqueitemsValidation1.getInstance()
        assertFailsWith<ValidationException>(
            block = {
                schema.validate(
                    listOf(
                        1,
                        1
                    ),
                    configuration
                )
            }
        )
    }

    @Test
    fun testNested0AndFalseAreUniquePasses() {
        // nested [0] and [false] are unique
        val schema = UniqueitemsValidation.UniqueitemsValidation1.getInstance()
        schema.validate(
            listOf(
                listOf(
                    listOf(
                        0
                    ),
                    "foo"
                ),
                listOf(
                    listOf(
                        false
                    ),
                    "foo"
                )
            ),
            configuration
        )
    }

    @Test
    fun testObjectsAreNonUniqueDespiteKeyOrderFails() {
        // objects are non-unique despite key order
        val schema = UniqueitemsValidation.UniqueitemsValidation1.getInstance()
        assertFailsWith<ValidationException>(
            block = {
                schema.validate(
                    listOf(
                        mapOf(
                            Pair(
                                "a",
                                1
                            ),
                            Pair(
                                "b",
                                2
                            )
                        ),
                        mapOf(
                            Pair(
                                "b",
                                2
                            ),
                            Pair(
                                "a",
                                1
                            )
                        )
                    ),
                    configuration
                )
            }
        )
    }

    @Test
    fun testNonUniqueArrayOfArraysIsInvalidFails() {
        // non-unique array of arrays is invalid
        val schema = UniqueitemsValidation.UniqueitemsValidation1.getInstance()
        assertFailsWith<ValidationException>(
            block = {
                schema.validate(
                    listOf(
                        listOf(
                            "foo"
                        ),
                        listOf(
                            "foo"
                        )
                    ),
                    configuration
                )
            }
        )
    }

    @Test
    fun testAFalseAndA0AreUniquePasses() {
        // {\\\&quot;a\\\&quot;: false} and {\\\&quot;a\\\&quot;: 0} are unique
        val schema = UniqueitemsValidation.UniqueitemsValidation1.getInstance()
        schema.validate(
            listOf(
                mapOf(
                    Pair(
                        "a",
                        false
                    )
                ),
                mapOf(
                    Pair(
                        "a",
                        0
                    )
                )
            ),
            configuration
        )
    }

    @Test
    fun testNonUniqueArrayOfMoreThanTwoArraysIsInvalidFails() {
        // non-unique array of more than two arrays is invalid
        val schema = UniqueitemsValidation.UniqueitemsValidation1.getInstance()
        assertFailsWith<ValidationException>(
            block = {
                schema.validate(
                    listOf(
                        listOf(
                            "foo"
                        ),
                        listOf(
                            "bar"
                        ),
                        listOf(
                            "foo"
                        )
                    ),
                    configuration
                )
            }
        )
    }

    @Test
    fun test0AndFalseAreUniquePasses() {
        // [0] and [false] are unique
        val schema = UniqueitemsValidation.UniqueitemsValidation1.getInstance()
        schema.validate(
            listOf(
                listOf(
                    0
                ),
                listOf(
                    false
                )
            ),
            configuration
        )
    }

    @Test
    fun testNonUniqueArrayOfNestedObjectsIsInvalidFails() {
        // non-unique array of nested objects is invalid
        val schema = UniqueitemsValidation.UniqueitemsValidation1.getInstance()
        assertFailsWith<ValidationException>(
            block = {
                schema.validate(
                    listOf(
                        mapOf(
                            Pair(
                                "foo",
                                mapOf(
                                    Pair(
                                        "bar",
                                        mapOf(
                                            Pair(
                                                "baz",
                                                true
                                            )
                                        )
                                    )
                                )
                            )
                        ),
                        mapOf(
                            Pair(
                                "foo",
                                mapOf(
                                    Pair(
                                        "bar",
                                        mapOf(
                                            Pair(
                                                "baz",
                                                true
                                            )
                                        )
                                    )
                                )
                            )
                        )
                    ),
                    configuration
                )
            }
        )
    }

    @Test
    fun testNumbersAreUniqueIfMathematicallyUnequalFails() {
        // numbers are unique if mathematically unequal
        val schema = UniqueitemsValidation.UniqueitemsValidation1.getInstance()
        assertFailsWith<ValidationException>(
            block = {
                schema.validate(
                    listOf(
                        1.0,
                        1.0,
                        1
                    ),
                    configuration
                )
            }
        )
    }

    @Test
    fun testNonUniqueArrayOfStringsIsInvalidFails() {
        // non-unique array of strings is invalid
        val schema = UniqueitemsValidation.UniqueitemsValidation1.getInstance()
        assertFailsWith<ValidationException>(
            block = {
                schema.validate(
                    listOf(
                        "foo",
                        "bar",
                        "foo"
                    ),
                    configuration
                )
            }
        )
    }

    @Test
    fun testUniqueArrayOfNestedObjectsIsValidPasses() {
        // unique array of nested objects is valid
        val schema = UniqueitemsValidation.UniqueitemsValidation1.getInstance()
        schema.validate(
            listOf(
                mapOf(
                    Pair(
                        "foo",
                        mapOf(
                            Pair(
                                "bar",
                                mapOf(
                                    Pair(
                                        "baz",
                                        true
                                    )
                                )
                            )
                        )
                    )
                ),
                mapOf(
                    Pair(
                        "foo",
                        mapOf(
                            Pair(
                                "bar",
                                mapOf(
                                    Pair(
                                        "baz",
                                        false
                                    )
                                )
                            )
                        )
                    )
                )
            ),
            configuration
        )
    }

    @Test
    fun testUniqueArrayOfArraysIsValidPasses() {
        // unique array of arrays is valid
        val schema = UniqueitemsValidation.UniqueitemsValidation1.getInstance()
        schema.validate(
            listOf(
                listOf(
                    "foo"
                ),
                listOf(
                    "bar"
                )
            ),
            configuration
        )
    }

    @Test
    fun testTrueIsNotEqualToOnePasses() {
        // true is not equal to one
        val schema = UniqueitemsValidation.UniqueitemsValidation1.getInstance()
        schema.validate(
            listOf(
                1,
                true
            ),
            configuration
        )
    }

    @Test
    fun testNested1AndTrueAreUniquePasses() {
        // nested [1] and [true] are unique
        val schema = UniqueitemsValidation.UniqueitemsValidation1.getInstance()
        schema.validate(
            listOf(
                listOf(
                    listOf(
                        1
                    ),
                    "foo"
                ),
                listOf(
                    listOf(
                        true
                    ),
                    "foo"
                )
            ),
            configuration
        )
    }

    @Test
    fun testUniqueArrayOfStringsIsValidPasses() {
        // unique array of strings is valid
        val schema = UniqueitemsValidation.UniqueitemsValidation1.getInstance()
        schema.validate(
            listOf(
                "foo",
                "bar",
                "baz"
            ),
            configuration
        )
    }

    @Test
    fun testFalseIsNotEqualToZeroPasses() {
        // false is not equal to zero
        val schema = UniqueitemsValidation.UniqueitemsValidation1.getInstance()
        schema.validate(
            listOf(
                0,
                false
            ),
            configuration
        )
    }

    @Test
    fun testUniqueArrayOfIntegersIsValidPasses() {
        // unique array of integers is valid
        val schema = UniqueitemsValidation.UniqueitemsValidation1.getInstance()
        schema.validate(
            listOf(
                1,
                2
            ),
            configuration
        )
    }

    @Test
    fun testDifferentObjectsAreUniquePasses() {
        // different objects are unique
        val schema = UniqueitemsValidation.UniqueitemsValidation1.getInstance()
        schema.validate(
            listOf(
                mapOf(
                    Pair(
                        "a",
                        1
                    ),
                    Pair(
                        "b",
                        2
                    )
                ),
                mapOf(
                    Pair(
                        "a",
                        2
                    ),
                    Pair(
                        "b",
                        1
                    )
                )
            ),
            configuration
        )
    }

    @Test
    fun testUniqueHeterogeneousTypesAreValidPasses() {
        // unique heterogeneous types are valid
        val schema = UniqueitemsValidation.UniqueitemsValidation1.getInstance()
        schema.validate(
            listOf(
                mapOf<Any?, Any?>(
                ),
                listOf(
                    1
                ),
                true,
                null,
                1,
                "{}"
            ),
            configuration
        )
    }

    @Test
    fun testUniqueArrayOfObjectsIsValidPasses() {
        // unique array of objects is valid
        val schema = UniqueitemsValidation.UniqueitemsValidation1.getInstance()
        schema.validate(
            listOf(
                mapOf(
                    Pair(
                        "foo",
                        "bar"
                    )
                ),
                mapOf(
                    Pair(
                        "foo",
                        "baz"
                    )
                )
            ),
            configuration
        )
    }

    @Test
    fun testNonUniqueHeterogeneousTypesAreInvalidFails() {
        // non-unique heterogeneous types are invalid
        val schema = UniqueitemsValidation.UniqueitemsValidation1.getInstance()
        assertFailsWith<ValidationException>(
            block = {
                schema.validate(
                    listOf(
                        mapOf<Any?, Any?>(
                        ),
                        listOf(
                            1
                        ),
                        true,
                        null,
                        mapOf<Any?, Any?>(
                        ),
                        1
                    ),
                    configuration
                )
            }
        )
    }
}
